src/util/heap/chunk_map.rs

```diff
@@ -66,22 +114,40 @@ impl ChunkMap {
     pub const ALLOC_TABLE: SideMetadataSpec =
         crate::util::metadata::side_metadata::spec_defs::CHUNK_MARK;
 
-    pub fn new() -> Self {
+    pub fn new(space_index: usize) -> Self {
         Self {
+            space_index,
             chunk_range: Mutex::new(Chunk::ZERO..Chunk::ZERO),
         }
     }
 
-    /// Set chunk state
-    pub fn set(&self, chunk: Chunk, state: ChunkState) {
+    /// Set a chunk as allocated, or as free.
+    pub fn set_allocated(&self, chunk: Chunk, allocated: bool) {
+        let state = if allocated {
+            ChunkState::allocated(self.space_index)
+        } else {
+            ChunkState::free()
+        };
         // Do nothing if the chunk is already in the expected state.
-        if self.get(chunk) == state {
+        if self.get_internal(chunk) == state {
             return;
         }
+        #[cfg(debug_assertions)]
+        {
+            let old_state = self.get_internal(chunk);
+            // If a chunk is free, any space may use it. If a chunk is not free, only the current space may update its state.
+            assert!(
+                old_state.is_free() || old_state.get_space_index() == self.space_index,
+                "Chunk {:?}: old state {:?}, new state {:?}. Cannot set to new state.",
+                chunk,
+                old_state,
+                state
+            );
+        }
         // Update alloc byte
-        unsafe { Self::ALLOC_TABLE.store::<u8>(chunk.start(), state as u8) };
+        unsafe { Self::ALLOC_TABLE.store::<u8>(chunk.start(), state.0) };
         // If this is a newly allcoated chunk, then expand the chunk range.
-        if state == ChunkState::Allocated {
+        if state.is_allocated() {
```

Comment:
This can just be
```suggestion
        if allocated {
```
right?